src/gh_pr_upsert/git.py

```diff
@@ -0,0 +1,223 @@
+                "-f",
+                f"title={title}",
+                "-f",
+                f"body={body}",
+            ],
+            json=True,
+        )
+
+        return cls.from_json(base_repo, head_repo, head_branch, json)
+
+    @classmethod
+    @cache
+    def get(cls, base_repo, base_branch, head_repo, head_branch):
+        matching_prs = run(
+            [
+                "gh",
+                "api",
+                "--header",
+                "X-GitHub-Api-Version:2022-11-28",
+                "--paginate",
+                "--method",
+                "GET",
+                f"/repos/{base_repo.owner}/{base_repo.name}/pulls",
+                "-f",
+                f"base={base_branch}",
+                "-f",
+                f"head={head_repo.owner}:{head_branch}",
+                "-f",
+                "state=open",
+            ],
+            json=True,
+        )
+
+        if not matching_prs:
+            return None
+
+        assert len(matching_prs) == 1
+
+        json = matching_prs[0]
+
+        return cls.from_json(base_repo, head_repo, head_branch, json)
+
+    def close(self, comment) -> None:
+        run(
+            [
+                "gh",
+                "pr",
+                "close",
+                "--repo",
+                self.base_repo.name_with_owner,
+                "--delete-branch",
+                "--comment",
+                comment,
+                str(self.number),
+            ]
+        )
+
+
+@cache
+def branch_exists(remote: str, branch: str) -> bool:
+    """Return True if `remote` has a branch named `branch`."""
+    try:
+        run(["git", "show-ref", f"refs/remotes/{remote}/{branch}"])
+    except CalledProcessError as err:
+        if err.returncode == 1:
+            return False
+        raise
+
+    return True
+
+
+@cache
+def configured_user():
+    """Return the configured git user."""
+    return User(
+        name=run(["git", "config", "--get", "user.name"]),
+        email=run(["git", "config", "--get", "user.email"]),
+    )
+
+
+@cache
+def current_branch() -> str:
+    """Return the name of the current local git branch."""
+    return run(["git", "symbolic-ref", "--quiet", "--short", "HEAD"])
+
+
+@cache
+def diff(branches: list[str]) -> str:
+    """Return the output of `git diff <branch>...` for the given `branches`."""
+    return run(["git", "diff", *branches])
+
+
+@cache
+def log(branches: list[str]) -> list[Commit]:
+    """Return the commits from `git log <branch>...` for the given `branches`."""
+    return [
+        Commit.get(sha)
+        for sha in run(
+            ["git", "log", "--ignore-missing", *branches, "--format=%H"]
+        ).split()
+    ]
+
+
+def push(remote: str, local_branch: str, remote_branch: str) -> None:
+    """Force-push <local_branch> to <remote>/<remote_branch>."""
+    run(
+        ["git", "push", "--force-with-lease", remote, f"{local_branch}:{remote_branch}"]
+    )
```

Comment:
`--force-with-lease` is how we avoid race conditions. `core.py` has used commands like `git log` and `git diff` to decide whether it's necessary and safe to force-push the branch. The info printed by these local git commands is based on the local git repo's tracking information about the remote branches from when the repo was cloned (or the last time you ran `git fetch --prune`). That local tracking info might be out of date. `git push --force-with-lease` force-pushes the branch **but crashes if the local tracking info is out of date**.